zcash_primitives/src/legacy/keys.rs

```diff
@@ -0,0 +1,198 @@
+            .map(ExternalPrivKey)
+    }
+}
+
+/// A type representing a BIP-44 public key at the account path level
+/// `m/44'/<coin_type>'/<account>'
+#[derive(Clone, Debug)]
+pub struct AccountPubKey(ExtendedPubKey);
+
+impl AccountPubKey {
+    /// Derive BIP-44 public key at the external child path
+    /// `m/44'/<coin_type>'/<account>'/0/<child_index>
+    pub fn derive_external_pubkey(
+        &self,
+        child_index: u32,
+    ) -> Result<ExternalPubKey, hdwallet::error::Error> {
+        self.0
+            .derive_public_key(KeyIndex::Normal(0))?
+            .derive_public_key(KeyIndex::Normal(child_index))
+            .map(ExternalPubKey)
+    }
+
+    pub fn from_extended_pubkey(extpubkey: ExtendedPubKey) -> Self {
+        AccountPubKey(extpubkey)
+    }
+
+    pub fn extended_pubkey(&self) -> &ExtendedPubKey {
+        &self.0
+    }
+}
+
+/// A type representing a private key at the BIP-44 external child
+/// level `m/44'/<coin_type>'/<account>'/0/<child_index>
+#[derive(Clone, Debug)]
+pub struct ExternalPrivKey(ExtendedPrivKey);
+
+impl ExternalPrivKey {
+    /// Returns the external public key corresponding to this private key
+    pub fn to_external_pubkey(&self) -> ExternalPubKey {
+        ExternalPubKey(ExtendedPubKey::from_private_key(&self.0))
+    }
+
+    /// Extracts the secp256k1 secret key component
+    pub fn secret_key(&self) -> &secp256k1::key::SecretKey {
+        &self.0.private_key
+    }
+}
+
+pub fn pubkey_to_address(pubkey: &secp256k1::key::PublicKey) -> TransparentAddress {
+    let mut hash160 = ripemd160::Ripemd160::new();
+    hash160.update(Sha256::digest(&pubkey.serialize()));
+    TransparentAddress::PublicKey(*hash160.finalize().as_ref())
+}
+
+/// A type representing a public key at the BIP-44 external child
+/// level `m/44'/<coin_type>'/<account>'/0/<child_index>
+#[derive(Clone, Debug)]
+pub struct ExternalPubKey(ExtendedPubKey);
+
+impl std::convert::TryFrom<&[u8; 65]> for ExternalPubKey {
+    type Error = hdwallet::error::Error;
+
+    fn try_from(data: &[u8; 65]) -> Result<Self, Self::Error> {
+        ExternalPubKey::deserialize(data)
+    }
+}
+
+impl ExternalPubKey {
+    /// Returns the transparent address corresponding to
+    /// this public key.
+    pub fn to_address(&self) -> TransparentAddress {
+        pubkey_to_address(&self.0.public_key)
+    }
+
+    /// Returns the secp256k1::key::PublicKey component of
+    /// this public key.
+    pub fn public_key(&self) -> &secp256k1::key::PublicKey {
+        &self.0.public_key
+    }
+
+    /// Returns the chain code component of this public key.
+    pub fn chain_code(&self) -> &[u8] {
+        &self.0.chain_code
+    }
+
+    /// Derives the internal ovk and external ovk corresponding to this
+    /// transparent fvk. As specified in [ZIP 316][transparent-ovk].
+    ///
+    /// [transparent-ovk]: https://zips.z.cash/zip-0316#deriving-internal-keys
+    pub fn ovks_for_shielding(&self) -> (InternalOvk, ExternalOvk) {
```

Comment:
This is a bug that was missed in the review of https://github.com/zcash/librustzcash/pull/480 before it was merged into this PR (I noted the bug after the fact here: https://github.com/zcash/librustzcash/pull/480#pullrequestreview-863948298). This should be defined on `AccountPubKey` instead, which is what becomes the transparent FVK.